### PR TITLE
ijar: fix manifest sections handling

### LINKED ISSUE #12730: ijar incorrectly handles MANIFESTs with sections
### Description of the problem:

ijar tool incorrectly handles MANIFEST files with [sections](https://docs.oracle.com/en/java/javase/11/docs/specs/jar/jar.html#name-value-pairs-and-sections) by removing empty lines and appending main attribute `Target-Label` after them.
I've faced this issue by trying to use `java_common.stamp_jar` on external jars.

### Bugs: what's the simplest, easiest way to reproduce this bug? Please provide a minimal example if possible.

Repro: https://github.com/liucijus/ijar-stamping-repro

### What operating system are you running Bazel on?

Linux (Xubuntu 20.04)

### What's the output of `bazel info release`?

release 3.7.2

### PR DESCRIPTION
**Pull request by liucijus**
Fixes #12730 

Important changes:
 - removed line stripping from the original implementation
 - label/rule attributes are appended after main attributes, then the rest of the `MANIFEST.MF` content is appended - this way sections data is preserved.

### CODE REVIEW

**Comment by ittaiz** [comment]
@cushon wdyt? Rather small PR which could help us out at rules_scala

**Comment by ittaiz** [comment]
Can anyone help with this? @liucijus stepped up and not only reported the issue but also sent a small PR and it's just sitting here.
FYI we're copy ~25k LOC into rules scala to be unblocked.
Divergences are happening :(

**Comment by ittaiz** [comment]
@aiuto process wise and bazel community- how long should a community contributor wait before getting a review? Especially for a small and confined PR?
I'm asking because this is demoralizing and works against your efforts of having not only Google look out for Bazel

File: third_party/ijar/test/BUILD [hunk 1]
```diff
@@ -144,6 +144,14 @@ genrule(
     tools = ["//third_party/ijar"],
 )
 
+genrule(
+    name = "jar_with_manifest_sections",
+    srcs = ["jar-with-manifest-sections.jar"],
```
Comment (comius) [inline third_party/ijar/test/BUILD:149]: For reference:
```
Archive:  jar-with-manifest-sections.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
     1382  2010-01-01 05:00   AnnotatedClass.class
      476  2010-01-01 05:00   Annotations.class
      381  2010-01-01 05:00   Annotations$ParametersOnlyAnnotation.class
      468  2010-01-01 05:00   Annotations$RuntimeInvisible.class
      561  2010-01-01 05:00   Annotations$RuntimeVisible.class
        0  2010-01-01 05:00   META-INF/
       96  2021-01-04 09:23   META-INF/MANIFEST.MF
       19  2018-03-29 15:53   textfile.txt
---------                     -------
     3383                     8 files
```

```
$ cat META-INF/MANIFEST.MF 
Manifest-Version: 1.0
Created-By: test-code

Name: foo
Foo: bar

Name: baz
Another: bar
```

**Review by comius — CHANGES_REQUESTED** [review]
LGTM, except that I would like to also have a test where Target-Label is in the manifest with sections and it is skipped.

**Comment by comius** [comment]
@cushon, do you think this could have any impact on the other Java code? My best guess is that it doesn't, because before the patch manifests with sections would just get mangled.

File: third_party/ijar/test/BUILD [hunk 2]
```diff
@@ -144,6 +144,22 @@ genrule(
...
+    outs = ["jar-with-manifest-sections-nostrip.jar"],
+    cmd = "$(location //third_party/ijar) --target_label //foo:foo --nostrip_jar $< $@",
+    tools = ["//third_party/ijar"],
+)
+
+genrule(
+    name = "jar_with_target_label_and_manifest_sections",
+    srcs = ["jar-with-target-label-and-manifest-sections.jar"],
```
Comment (comius) [inline third_party/ijar/test/BUILD:157]: For reference:
```
Archive:  jar-with-target-label-and-manifest-sections.jar
  Length      Date    Time    Name
---------  ---------- -----   ----
     1382  2010-01-01 05:00   AnnotatedClass.class
      476  2010-01-01 05:00   Annotations.class
      381  2010-01-01 05:00   Annotations$ParametersOnlyAnnotation.class
      468  2010-01-01 05:00   Annotations$RuntimeInvisible.class
      561  2010-01-01 05:00   Annotations$RuntimeVisible.class
        0  2010-01-01 05:00   META-INF/
      122  2021-01-29 08:19   META-INF/MANIFEST.MF
       19  2018-03-29 15:53   textfile.txt
---------                     -------
     3409                     8 files
```

```
$ cat META-INF/MANIFEST.MF 
Manifest-Version: 1.0
Created-By: test-code
Target-Label: //not:this

Name: foo
Foo: bar

Name: baz
Another: bar
```

**Review by comius — APPROVED** [review]

**Comment by comius** [comment]
Internal linter says: ".jar files are no longer allowed in third_party." and prevents me from adding new .jar file (while old ones are fine). Bazel should be able to generate those .jars from sources, but it's another curve ball.

**Comment by liucijus** [comment]
> LGTM, except that I would like to also have a test where Target-Label is in the manifest with sections and it is skipped.

I assume you want a test to show that existing label is update, right. If I'm not missing anything, it's the way ijar works.

**Comment by comius** [comment]
> > LGTM, except that I would like to also have a test where Target-Label is in the manifest with sections and it is skipped.
> 
> I assume you want a test to show that existing label is update, right? If I'm not missing anything, it's the way ijar works.

Correct. I already reviewed it, it's ok.

**Comment by comius** [comment]
> Internal linter says: ".jar files are no longer allowed in third_party." and prevents me from adding new .jar file (while old ones are fine). Bazel should be able to generate those .jars from sources, but it's another curve ball.

Do you have resource/will to fix this as well?

**Comment by liucijus** [comment]
> > Internal linter says: ".jar files are no longer allowed in third_party." and prevents me from adding new .jar file (while old ones are fine). Bazel should be able to generate those .jars from sources, but it's another curve ball.
> Do you have resource/will to fix this as well?

I'll look into it

**Comment by liucijus** [comment]
> > > Internal linter says: ".jar files are no longer allowed in third_party." and prevents me from adding new .jar file (while old ones are fine). Bazel should be able to generate those .jars from sources, but it's another curve ball.
> > > Do you have resource/will to fix this as well?
> 
> I'll look into it

Done

**Comment by ittaiz** [comment]
Thanks @liucijus and @comius!

**Pull request closed without merging**